### ISSUE TITLE
Document node reuse

### ISSUE THREAD

**Issue by rainersigwald** [Microsoft/msbuild#2916]
I searched docs.microsoft.com for details about node reuse so I could send them to a person who was asking. But there was basically nothing, and a search for [msbuild node reuse] wasn't super helpful, either.

There's an announcement blog post: https://blogs.msdn.microsoft.com/msbuild/2007/04/16/node-reuse-in-multiproc-msbuild/

But since it's important to tell people with CI servers and so on to disable node reuse, we should have clear documentation in the usual place.

**Comment by rainersigwald** [comment]
Sent this mail, which can serve as a start:

When you build a multiprocess build (with the `/m` switch to msbuild.exe), MSBuild starts “worker nodes” in the background that by default do not terminate when the build completes. On a subsequent build, MSBuild searches for idle worker nodes and connects to them, saving startup overhead and JIT cost. Since .NET doesn’t offer great ways to unload DLLs, those workers can then hold locks on task DLLs until they hit their idle shutdown.

There are a couple of ways to deal with this.

You can force this specific task to run in its own short-lived process by [setting `TaskFactory="TaskHostFactory"` in the UsingTask](https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-configure-targets-and-tasks#task-factories). That is often wise for tasks built in the same build process, but it does have some overhead (for the IPC cost of remoting the execution, plus process startup and JIT in the task host.

You can disable node reuse with the command line switch `/nodeReuse:false` or with the environment variable `MSBUILDDISABLENODEREUSE=1.`

**Comment by daveaglick** [comment]
> You can disable node reuse with the command line switch `/nodeReuse:false` or with the environment variable `MSBUILDDISABLENODEREUSE=1`.

Having just searched for docs regarding node reuse, this issue was helpful. Quick question: does disabling node reuse with the above methods terminate the MSBuild node upon completion, force spawning a new node at the start of a build, or both? I.e., if node reuse is turned off with the above, will MSBuild still use an existing node if one is found?

**Comment by daveaglick** [comment]
Never mind, just ran across https://github.com/Microsoft/msbuild/issues/339 which explains the situation nicely. I was having some problems with it reusing nodes even after setting `MSBUILDDISABLENODEREUSE` but it looks like that's because I still had a node around from before I made the change with similar handshake properties.

I would suggest the specifics of this behavior be added to any forthcoming documentation.

**Comment by rainersigwald** [comment]
@daveaglick also related: https://github.com/Microsoft/msbuild/pull/3158, getting us to both terminate-at-end-of-build and guarantee-started-fresh.

**Comment by chipplyman** [comment]
I'd love to see some figures on how much time is likely to be saved by reusing nodes.  I suspect it's in the sub-second range and thus only has an appreciable impact on small solutions containing only project types with excellent no-op build performance.  (e.g. IME .sqlprojs have terrible no-op characteristics)

**Comment by rainersigwald** [comment]
@Zastai TaskHostFactory has been in MSBuild since version 4.something (Visual Studio 2012). It was enabled for .NET (Core) SDK dotnet/msbuild#6994 for MSBuild 17.1/.NET SDK 6.0.200.

> is there an MSBuild variable indicating the status of node reuse

No, I don't think so.

> or multi-process builds

There's `$(MSBuildNodeCount)`.

> Any chance of getting a feature-test function, perhaps, like `Condition=" $([MSBuild]::HasFeature('task-host-factory')) "`. Feels cleaner than a version check.

Indeed, that's strongly preferred when testing for Windows features and I see no reason not to do it (other than "we could spend that time doing something else" which is always a pretty good reason). Feel free to file an issue.

**Comment by Zastai** [comment]
Right, so either msbuild type 'Full', or 'Core' with version >= 17.1. Will have to investigate when Mono MSBuild got it.
Doable, if not very nice. Will file an issue for the feature test.

**Comment by Zastai** [comment]
Small separate question: is there an MSBuild.exe equivalent of `dotnet build-server shutdown` (which I assume finds and terminates all applicable nodes)?

**Comment by rainersigwald** [comment]
> is there an MSBuild.exe equivalent of `dotnet build-server shutdown` (which I assume finds and terminates all applicable nodes)?

There is no `msbuild.exe` equivalent command-line call to the API that does that (today).

**Comment by baronfel** [comment]
Having said that, I would like to hook MSBuild up to that command when we complete the new MSBuild Server Daemon work and get that incorporated into the SDK.

**Comment by rainersigwald** [comment]
MSBuild-on-Core is already hooked up to `dotnet build-server shutdown`, but there's no alternative to `taskkill /f /im msbuild.exe` for MSBuild-on-Framework.
